core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -0,0 +1,246 @@
+            return location;
+        }
+    }
+
+    private class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult {
+        private final ReplicatedWriteRequest<?> request;
+        volatile boolean finishedWrite;
+        volatile ActionListener<Response> listener = null;
+
+        public WritePrimaryResult(Request request, Response finalResponse,
+                                  @Nullable Translog.Location location,
+                                  IndexShard indexShard) {
+            super(request, finalResponse);
+            this.request = request;
+            finishWrite(indexShard, request, location);
+        }
+
+        @Override
+        public synchronized void respond(ActionListener<Response> listener) {
+            this.listener = listener;
+            respondIfPossible();
+        }
+
+        /**
+         * Respond if the refresh has occurred and the listener is ready. Always called while synchronized on {@code this}.
+         */
+        protected void respondIfPossible() {
+            if (finishedWrite && listener != null) {
+                super.respond(listener);
+            }
+        }
+
+        @Override
+        public void forcedRefresh() {
+            finalResponse.setForcedRefresh(true);
+        }
+
+        @Override
+        public void waitForRefreshForcedRefresh() {
+            logger.warn("block_until_refresh request ran out of slots and forced a refresh: [{}]", request);
+        }
+
+        @Override
+        public synchronized void finished() {
+            finishedWrite = true;
+            respondIfPossible();
+        }
+    }
+
+    private class WriteReplicaResult extends ReplicaResult implements RespondingWriteResult {
+        private final IndexShard indexShard;
+        private final ReplicatedWriteRequest<?> request;
+        private final Translog.Location location;
+        private ActionListener<TransportResponse.Empty> listener;
+
+        public WriteReplicaResult(IndexShard indexShard, ReplicatedWriteRequest<?> request, Translog.Location location) {
+            this.indexShard = indexShard;
+            this.request = request;
+            this.location = location;
+        }
+
+        @Override
+        public void respond(ActionListener<TransportResponse.Empty> listener) {
+            this.listener = listener;
+            finishWrite(indexShard, request, location);
+        }
+
+        @Override
+        public void forcedRefresh() {
+            // It'd be nice to mark this in the response but logging when wait for refresh forced it is OK.
+        }
+
+        @Override
+        public void waitForRefreshForcedRefresh() {
+            logger.warn("block_until_refresh request ran out of slots and forced a refresh: [{}]", request);
+        }
+
+        @Override
+        public void finished() {
+            listener.onResponse(TransportResponse.Empty.INSTANCE);
+        }
+    }
+
+    private interface RespondingWriteResult {
```

Comment:
@bleskes: this is how I removed the duplication between the primary and the replica.